include/irr/video/IGPUVirtualTexture.h

```diff
@@ -0,0 +1,188 @@
+            base_t(
+                createGPUImageFromCPU(_driver, _am, _cpuStorage->image.get()),
+                _cpuStorage->m_assignedPageTableLayers,
+                _cpuStorage->m_addr_layerShift,
+                _cpuStorage->m_addr_xMask
+            ),
+            m_driver(_driver)
+        {
+
+        }
+
+        IGPUVTResidentStorage(IVideoDriver* _driver, asset::E_FORMAT _format, uint32_t _extent, uint32_t _layers, uint32_t _tilesPerDim) :
+            base_t(_layers, _tilesPerDim),
+            m_driver(_driver)
+        {
+            IGPUImage::SCreationParams params;
+            params.extent = { _extent,_extent,1u };
+            params.format = _format;
+            params.arrayLayers = _layers;
+            params.mipLevels = 1u;
+            params.type = asset::IImage::ET_2D;
+            params.samples = asset::IImage::ESCF_1_BIT;
+            params.flags = static_cast<asset::IImage::E_CREATE_FLAGS>(0);
+
+            image = m_driver->createDeviceLocalGPUImageOnDedMem(std::move(params));
+        }
+
+    private:
+        core::smart_refctd_ptr<IGPUImageView> createView_internal(IGPUImageView::SCreationParams&& _params) const override
+        {
+            return m_driver->createGPUImageView(std::move(_params));
+        }
+
+        IVideoDriver* m_driver = nullptr;
+    };
+
+public:
+    IGPUVirtualTexture(
+        IVideoDriver* _driver,
+        const base_t::IVTResidentStorage::SCreationParams* _residentStorageParams,
+        uint32_t _residentStorageCount,
+        uint32_t _pgTabSzxy_log2 = 8u,
+        uint32_t _pgTabLayers = 256u,
+        uint32_t _pgSzxy_log2 = 7u,
+        uint32_t _tilePadding = 9u,
+        uint32_t _maxAllocatableTexSz_log2 = 14u
+    ) :
+        base_t(
+            _pgTabSzxy_log2,
+            _pgTabLayers,
+            _pgSzxy_log2,
+            _tilePadding
+        ),
+        m_driver(_driver)
+    {
+        m_pageTable = createPageTable(_pgTabSzxy_log2, _pgTabLayers, _pgSzxy_log2, _maxAllocatableTexSz_log2);
+        initResidentStorage(_residentStorageParams, _residentStorageCount);
+    }
+    IGPUVirtualTexture(IVideoDriver* _driver, asset::IAssetManager* _am, asset::ICPUVirtualTexture* _cpuvt) :
+        base_t(
+            _cpuvt->getPageExtent_log2(),
+            _cpuvt->getPageTable()->getCreationParameters().arrayLayers,
+            _cpuvt->getPageExtent_log2(),
+            _cpuvt->getTilePadding(),
+            false
+            ),
+        m_driver(_driver)
+    {
+        //now copy from CPU counterpart resources that can be shared (i.e. just copy state) between CPU and GPU
+        //and convert to GPU those which can't be "shared": page table and VT resident storages along with their images and views
+
+        m_pageTable = std::move(_driver->getGPUObjectsFromAssets(&_cpuvt->getPageTable(), &_cpuvt->getPageTable()+1)->front());
+
+        m_freePageTableLayerIDs = _cpuvt->getFreePageTableLayersStack();
```

Comment:
why on earth are you hooking it up like that?

Reply:
ok I see, you're copying.... but that function shouldn't be public, make it protected in the `IVirtualTexture`

Reply:
because i dont really see other way

Reply:
protected wont do because IVirtualTexture is template and protected in IVirtualTexture<ICPUImageView> doesnt make it accessible in IGPUVirtualTexture with base being IVirtualTexture<IGPUImageView>

Reply:
this function returns const reference so it's safe, not the most beutiful thing in the world though

Reply:
eh, a stack doesn't make sense anyway ;)